### PR TITLE
report: show source line info with --srcline option

### PR DESCRIPTION
**Pull request by ekyooo**
Show source line info of each function with loaded debug_info.

Fixed: #763

Could I leave 'Signed off' message after code review?

I think there are many other ways to implement this feature.
All comments are welcome.

### CODE REVIEW

**Comment by ekyooo** [comment]
These are test results.


```
$ uftrace record --srcline t-abc
$ uftrace report --srcline
  Total time   Self time       Calls  Function [Source]
  ==========  ==========  ==========  ====================
   17.508 us    2.199 us           1  main [./tests/s-abc.c:26]
   15.309 us    2.384 us           1  a [./tests/s-abc.c:11]
   12.925 us    2.633 us           1  b [./tests/s-abc.c:16]
   10.292 us    5.159 us           1  c [./tests/s-abc.c:21]
    5.133 us    5.133 us           1  getpid
    3.437 us    3.437 us           1  __monstartup
    1.959 us    1.959 us           1  __cxa_atexit


$ .uftrace record --srcline test/t-lib
$ uftrace report --srcline
  Total time   Self time       Calls  Function [Source]
  ==========  ==========  ==========  ====================
   62.319 us    6.244 us           3  lib_a
   31.980 us    3.679 us           2  lib_b
   27.758 us    1.994 us           1  main [/home/xxx/soft/uftrace/tests/s-libmain.c:16]
   25.764 us    2.687 us           1  foo [/home/xxx/soft/uftrace/tests/s-libmain.c:11]
   21.022 us   13.154 us           2  lib_c


Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
024 report_basic        : OK OK OK OK OK OK OK OK OK OK
025 report_s_call       : OK OK OK OK OK OK OK OK OK OK
044 report_avg_total    : OK OK OK OK OK OK OK OK OK OK
045 report_avg_self     : OK OK OK OK OK OK OK OK OK OK
046 report_task         : OK OK OK OK OK OK OK OK OK OK
067 report_diff         : OK OK OK OK OK OK OK OK OK OK
090 report_recursive    : OK OK OK OK OK OK OK OK OK OK
092 report_tid          : OK OK OK OK OK OK OK OK OK OK
093 report_filter       : OK OK OK OK OK OK OK OK OK OK
094 report_depth        : OK OK OK OK OK OK OK OK OK OK
106 report_time         : OK OK OK OK OK OK OK OK OK OK
125 report_range        : OK OK OK OK OK OK OK OK OK OK
158 report_diff_policy1 : OK OK OK OK OK OK OK OK OK OK
159 report_diff_policy2 : OK OK OK OK OK OK OK OK OK OK
160 report_diff_policy3 : OK OK OK OK OK OK OK OK OK OK
164 report_sched        : SK SK SK SK SK SK SK SK SK SK
177 report_diff_policy4 : OK OK OK OK OK OK OK OK OK OK
216 no_libcall_report   : OK OK OK OK OK OK OK OK OK OK
```

File: doc/uftrace-report.md [hunk 1]
```diff
@@ -58,6 +58,9 @@ REPORT OPTIONS
     `--data` option, index 1 is for data given by the `--diff` option, and index
     2 is for (percentage) differences between the two data.
 
+\--srcline
+:   Show source line of each function that include debug information in object.
```
Comment (honggyukim) [inline doc/uftrace-report.md]: How about this?
```suggestion
:   Show source line of each function if available.
```

File: doc/uftrace-report.md [hunk 2]
```diff
@@ -152,6 +155,17 @@ This command shows information like the following:
        26.046 us    0.939 us           1  a
        25.107 us    0.934 us           1  b
 
+    $ uftrace report --srcline
```
Comment (honggyukim) [inline doc/uftrace-report.md]: This example actually doesn't show source location because the target program is not recorded with `--srcline` and `abc` is may not be compiled with `-g`.

So I think we better write the example as follows:
```suggestion
    $ uftrace record --srcline abc
    $ uftrace report --srcline
```

**Comment by namhyung** [comment]
> Could I leave 'Signed off' message after code review?

Yes, please.  The sign-off process is to ensure it's ok for open source as either you wrote the code or you know it's from someone who wrote the code to be open-sourced.

File: cmds/report.c [hunk 3]
```diff
@@ -225,6 +236,9 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)
 	struct rb_root sort_root = RB_ROOT;
 	const char f_format[] = "  %10.10s  %10.10s  %10.10s  %-.*s\n";
 	const char line[] = "=================================================";
+	char field_func_name[20] = "Function";
+	if (opts->srcline)
+		strcat(field_func_name, " [Source]");
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply cmds/report.c]: Ok. Thank you.

File: doc/uftrace-report.md (hunk 1 above)
Comment (ekyooo) [reply doc/uftrace-report.md]: Ok. Thank you.

File: doc/uftrace-report.md (hunk 2 above)
Comment (ekyooo) [reply doc/uftrace-report.md]: I agree. Thank you.

File: cmds/report.c [hunk 4]
```diff
@@ -194,7 +199,10 @@ static void print_function(struct uftrace_report_node *node, void *unused)
 		print_time_unit(node->total.sum);
 		pr_out("  ");
 		print_time_unit(node->self.sum);
-		pr_out("  %10"PRIu64 "  %-s\n", node->call, node->name);
+		pr_out("  %10"PRIu64 "  %-s", node->call, node->name);
+		if (node->loc)
```
[the remark this answers is not in the document]
Comment (ekyooo) [reply cmds/report.c]: Thank you for finding and debugging this issue.
I intended task_find_sym_addr to return valid debug_location or NULL. So I added additional check to task_find_sym_addr function.

I found that some symbols described in .debug_info like below is not collected to *.dbg file though it seems like source location info. 
I'll try to know the rule when available, but if you can simply share the rule of collecting debug location from dwarf to *.dbg, It'll be helpful.
Thank you.
```
 6846068  <3><e89af4>: Abbrev Number: 398 (DW_TAG_subprogram)
 6846069     <e89af6>   DW_AT_name        : (indirect string, offset: 0x420b0a): PlatformWorkerThread 
 6846070     <e89afa>   DW_AT_decl_file   : 3
 6846071     <e89afb>   DW_AT_decl_line   : 29
 6846072     <e89afc>   DW_AT_declaration : 1
                                      // no DW_AT_linkage_name
```

File: utils/session.c [hunk 5]
```diff
@@ -655,6 +655,57 @@ struct sym * task_find_sym_addr(struct uftrace_session_link *sessions,
...
+
+		if (is_kernel_address(&fsess->symtabs, addr))
+			sess = fsess;
+		else
+			return NULL;
+	}
+
+	sym = find_symtabs(&sess->symtabs, addr);
+	if (sym == NULL)
+		sym = session_find_dlsym(sess, time, addr);
```
Comment (ekyooo) [inline utils/session.c]: Can sym actually be null?
if sym can be null, I need to check it in next line before accessing it.

**Comment by honggyukim** [comment]
The same change can also be applied to replay command as well. I look forward to having this feature.

**Comment by ekyooo** [comment]
Does "this feature" mean replaying with srcline?
If so, I will gladly do it after the review of this patch is complete.
Are you waiting for something from me for this patch except rebasing?

**Comment by honggyukim** [comment]
> Does "this feature" mean replaying with srcline?

I mean both report and replay.

> Are you waiting for something from me for this patch except rebasing?

This PR needs to be rebased.

File: cmds/report.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply cmds/report.c:209]: Sorry for late reply.  The basic rule is that it checks DW_TAG_subprogram DIE and reads decl_file and decl_line attributes.  It should have the location info if *.dbg file exists.

File: utils/session.c (hunk 5 above)
Comment (namhyung) [reply utils/session.c:689]: If addr is not found in the symbol table, it can be NULL.

File: utils/session.c [hunk 6]
```diff
@@ -655,6 +655,60 @@ struct sym * task_find_sym_addr(struct uftrace_session_link *sessions,
...
+		if (map == NULL)
+			return NULL;
+
+		struct debug_info *dinfo = &(map->mod->dinfo);
+		if (dinfo == NULL || dinfo->nr_locs_used == 0)
+			return NULL;
+
+		ptrdiff_t sym_idx = sym - map->mod->symtab.sym;
+		struct debug_location *loc = &dinfo->locs[sym_idx];
+		if (loc->file != 0)
```
Comment (namhyung) [inline utils/session.c:705]: Please use NULL for pointer comparision.

File: cmds/report.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply cmds/report.c:209]: @honggyukim The `task_find_loc_addr()` already checked the file pointer so I think it's ok.

File: utils/session.c (hunk 6 above)
Comment (ekyooo) [reply utils/session.c]: Thank you. It's fixed.

File: utils/session.c (hunk 5 above)
Comment (ekyooo) [reply utils/session.c:689]: Thank you for your information.
It checked in line 691 to return NULL when it's NULL.

File: cmds/report.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ekyooo) [reply cmds/report.c:209]: > Sorry for late reply. The basic rule is that it checks DW_TAG_subprogram DIE and reads decl_file and decl_line attributes. It should have the location info if *.dbg file exists.

Then do you think it would be an issue?

**Comment by ekyooo** [comment]
> Thanks for working on this. Could you please add a test case?

I missed this comment. 
Sure, I'll update. 
Thank you.

File: cmds/report.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply cmds/report.c:209]: No, I think it's ok now.

File: cmds/report.c [hunk 7]
```diff
@@ -569,6 +590,9 @@ static void report_diff(struct uftrace_data *handle, struct opts *opts)
 	};
 	int h_idx = (avg_mode == AVG_NONE) ? 0 : (avg_mode == AVG_TOTAL) ? 1 : 2;
 	int f_idx = diff_policy.percent ? 1 : (avg_mode == AVG_NONE) ? 0 : 2;
+	char field_func_name[] = "Function [Source]";
+	if (!opts->srcline)
```
Comment (honggyukim) [inline cmds/report.c:594]: Please give an empty line after declaration statements.

File: utils/session.c [hunk 8]
```diff
@@ -655,6 +655,60 @@ struct sym * task_find_sym_addr(struct uftrace_session_link *sessions,
...
+		map = find_map(&sess->symtabs, addr);
+		if (map == NULL)
+			return NULL;
+
+		struct debug_info *dinfo = &(map->mod->dinfo);
+		if (dinfo == NULL || dinfo->nr_locs_used == 0)
+			return NULL;
+
+		ptrdiff_t sym_idx = sym - map->mod->symtab.sym;
+		struct debug_location *loc = &dinfo->locs[sym_idx];
```
Comment (honggyukim) [inline utils/session.c:704]: uftrace code style recommends placing all the declarations at the top of each block. Could you please split the declarations and assignments?

File: utils/report.c [hunk 9]
```diff
@@ -113,10 +115,10 @@ void report_update_node(struct uftrace_report_node *node,
 
 	total_time = fstack->total_time;
 	self_time = fstack->total_time - fstack->child_time;
-
```
Comment (honggyukim) [inline utils/report.c:116]: Can you keep this unnecessary change?

File: cmds/report.c (hunk 7 above)
Comment (ekyooo) [reply cmds/report.c:594]: ok.

File: utils/report.c (hunk 9 above)
Comment (ekyooo) [reply utils/report.c:116]: ok.

File: utils/session.c (hunk 8 above)
Comment (ekyooo) [reply utils/session.c:704]: ok.

File: cmds/report.c (hunk 7 above)
Comment (honggyukim) [reply cmds/report.c:595]: It's not updated yet.

File: utils/report.c [hunk 10]
```diff
@@ -512,7 +515,7 @@ void report_diff_nodes(struct rb_root *orig_root, struct rb_root *pair_root,
 void destroy_diff_nodes(struct rb_root *diff_root)
 {
 	struct rb_node *n = rb_first(diff_root);
-	
+
```
Comment (honggyukim) [inline utils/report.c:518]: please do not change here.

**Comment by ekyooo** [comment]
FYI, this is test results new new tc.
```
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
235 report_srcline      : OK OK OK OK OK SK SK SK SK SK
```

@honggyukim 
Did you finish the review?
I'm going to update after I get all reviews including test case at once.

> In addition, could you please update your name with `git config --global user.name "Your Name"`

Want to get rid of anonymous freedom? :)
Can you let me know why?

**Comment by ekyooo** [comment]
> The same change can also be applied to replay command as well. 

For replay, do you want to include srcline info in the Funciotn field like this patch, or to handle it as a field? If it's need to discuss about details about it, could you open an issue or give any opinion to progress this?

File: utils/report.c (hunk 10 above)
Comment (ekyooo) [reply utils/report.c:518]: Original code has just space in empty line like below. it's from master branch.
Do you really want to keep this space?

```
514 >-------struct rb_node *n = rb_first(diff_root);$
515 >-------$
516 >-------while (n) {$
```

Comment (namhyung) [reply utils/report.c:518]: No, please remove the spaces.

**Comment by namhyung** [comment]
@honggyukim do you still have some concerns with this PR?  I think @ekyooo addressed all your comments.

**Comment by honggyukim** [comment]
> For replay, do you want to include srcline info in the Funciotn field like this patch, or to handle it as a field? If it's need to discuss about details about it, could you open an issue or give any opinion to progress this?

I think it can be handled in the same way by placing "Source" info right next to the function names. I don't mean to add it as a field. You can create an issue with a very simple description. It doesn' have to be long, but just one line is also enough, then add more comments incrementally.

So feel free to create an issue when needed. If you want me to create it, then I will do that as well.

**Comment by ekyooo** [comment]
> As a follow-up step, it'd be nice if we could remove the common directory prefix if file name is under the compile unit's compilation directory (`DW_AT_comp_dir`).
> 
> So the srcline output could have "foobar.c" part only.

Is it ok to save "foobar.c" (in above example) to *.dbg file in that case?

**Comment by namhyung** [comment]
> Is it ok to save "foobar.c" (in above example) to *.dbg file in that case?

Right!

**Comment by honggyukim** [comment]
How about applying the following change on top of this PR?
```diff
diff --git a/cmds/report.c b/cmds/report.c
index 39a3edd..283ef39 100644
--- a/cmds/report.c
+++ b/cmds/report.c
@@ -207,7 +207,7 @@ static void print_function(struct uftrace_report_node *node, void *unused)
                print_time_unit(node->self.sum);
                pr_out("  %10"PRIu64 "  %-s", node->call, node->name);
                if (node->loc)
-                       pr_out(" [%s:%d]", node->loc->file->name, node->loc->line);
+                       pr_gray(" [%s:%d]", node->loc->file->name, node->loc->line);
                pr_out("\n");
        }
        else {
@@ -231,7 +231,7 @@ static void print_function(struct uftrace_report_node *node, void *unused)
                print_time_unit(time_max);
                pr_out("  %-s", node->name);
                if (node->loc)
-                       pr_out(" [%s:%d]", node->loc->file->name, node->loc->line);
+                       pr_gray(" [%s:%d]", node->loc->file->name, node->loc->line);
                pr_out("\n");
        }
 }
@@ -240,12 +240,8 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)
 {
        struct rb_root name_root = RB_ROOT;
        struct rb_root sort_root = RB_ROOT;
-       const char f_format[] = "  %10.10s  %10.10s  %10.10s  %-.*s\n";
+       const char f_format[] = "  %10.10s  %10.10s  %10.10s  %-.*s";
        const char line[] = "=================================================";
-       char field_func_name[] = "Function [Source]";
-
-       if (!opts->srcline)
-               field_func_name[8] = '\0';

        build_function_tree(handle, &name_root, opts);
        report_calc_avg(&name_root);
@@ -256,13 +252,17 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)


        if (avg_mode == AVG_NONE)
-               pr_out(f_format, "Total time", "Self time", "Calls", maxlen, field_func_name);
+               pr_out(f_format, "Total time", "Self time", "Calls", maxlen, "Function");
        else if (avg_mode == AVG_TOTAL)
-               pr_out(f_format, "Avg total", "Min total", "Max total", maxlen, field_func_name);
+               pr_out(f_format, "Avg total", "Min total", "Max total", maxlen, "Function");
        else if (avg_mode == AVG_SELF)
-               pr_out(f_format, "Avg self", "Min self", "Max self", maxlen, field_func_name);
+               pr_out(f_format, "Avg self", "Min self", "Max self", maxlen, "Function");
+       if (opts->srcline)
+               pr_gray(" [Source]");
+       pr_out("\n");

        pr_out(f_format, line, line, line, maxlen, line);
+       pr_out("\n");

        print_and_delete(&sort_root, true, NULL, print_function);
 }
```
It makes my eyes so much comfortable. :)

**Comment by ekyooo** [comment]
> How about applying the following change on top of this PR?

I think it's good.
Thank you for good opinion.
Do you mean you want me to apply that change?

**Comment by honggyukim** [comment]
Yes, please take this change with your PR.

**Comment by namhyung** [comment]
Also please check the alignment of examples in the commit log.

**Comment by honggyukim** [comment]
Please squash those two commits into one.

**Comment by ekyooo** [comment]
ok, I'll update it after I get confirmed you have no more requirement.

File: cmds/report.c [hunk 11]
```diff
@@ -239,14 +250,19 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)
...
+
 	if (avg_mode == AVG_NONE)
 		pr_out(f_format, "Total time", "Self time", "Calls", maxlen, "Function");
 	else if (avg_mode == AVG_TOTAL)
 		pr_out(f_format, "Avg total", "Min total", "Max total", maxlen, "Function");
 	else if (avg_mode == AVG_SELF)
 		pr_out(f_format, "Avg self", "Min self", "Max self", maxlen, "Function");
+	if (opts->srcline)
+		pr_out(" [Source]");
+	pr_out("\n");
```
Comment (honggyukim) [inline cmds/report.c:262]: please change this to `pr_gray`.

File: cmds/report.c [hunk 12]
```diff
@@ -239,14 +250,19 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)
 	if (uftrace_done)
 		return;
 
+
```
Comment (honggyukim) [inline cmds/report.c:253]: do not change this.

File: cmds/report.c [hunk 13]
```diff
@@ -240,12 +240,8 @@ static void report_functions(struct uftrace_data *handle, struct opts *opts)
 {
 	struct rb_root name_root = RB_ROOT;
 	struct rb_root sort_root = RB_ROOT;
-	const char f_format[] = "  %10.10s  %10.10s  %10.10s  %-.*s\n";
+	const char f_format[] = "  %10.10s  %10.10s  %10.10s  %-.*s";
 	const char line[] = "=================================================";
-	char field_func_name[] = "Function [Source]";
-
-	if (!opts->srcline)
-		field_func_name[8] = '\0';
```
Comment (namhyung) [inline cmds/report.c]: Please move this part into the previous commit so that we can skip unnecessary changes in the first place.

File: cmds/report.c [hunk 14]
```diff
@@ -591,10 +591,6 @@ static void report_diff(struct uftrace_data *handle, struct opts *opts)
 	};
 	int h_idx = (avg_mode == AVG_NONE) ? 0 : (avg_mode == AVG_TOTAL) ? 1 : 2;
 	int f_idx = diff_policy.percent ? 1 : (avg_mode == AVG_NONE) ? 0 : 2;
-	char field_func_name[] = "Function [Source]";
-
-	if (!opts->srcline)
-		field_func_name[8] = '\0';
```
Comment (namhyung) [inline cmds/report.c]: Ditto.

**Comment by ekyooo** [comment]
How about sharing your coding style on the page below as general project does? 
(for example, placing declarations, usage example on git commit msg, about spacing, and maybe more)
https://github.com/namhyung/uftrace/blob/master/CONTRIBUTING.md

It'd be good to check all your coding style for myself before first pushing rather than knowing one by one through repeated update.
Maybe it could save time on each other and focus on the logical part.
Thank you.

**Comment by namhyung** [comment]
Well, it has a small coding style section in the doc.  The idea is that we just follow Linux kernel coding style with an exception.  Please refer https://www.kernel.org/doc/html/v5.6/process/coding-style.html.

Also note that, many editors and IDEs support the Linux kernel codying style rules by default.

**Comment by namhyung** [comment]
There's an on-going effort to improve this situation: #1145
